core/src/runtime/mod.rs

```diff
@@ -92,22 +94,59 @@ pub struct Runtime {
     pub should_report: bool,
 }
 
+/// Integer type used to count instructions and syscalls for `ExecutionReport`.
+type Count = u64;
+
 #[derive(Default, Debug, Clone, PartialEq, Eq)]
 pub struct ExecutionReport {
-    pub instruction_counts: HashMap<Opcode, u64>,
-    pub syscall_counts: HashMap<SyscallCode, u64>,
+    pub instruction_counts: HashMap<Opcode, Count>,
+    pub syscall_counts: HashMap<SyscallCode, Count>,
 }
 
 impl ExecutionReport {
-    pub fn total_instruction_count(&self) -> u64 {
+    pub fn total_instruction_count(&self) -> Count {
         self.instruction_counts.values().sum()
     }
 
-    pub fn total_syscall_count(&self) -> u64 {
+    pub fn total_syscall_count(&self) -> Count {
         self.syscall_counts.values().sum()
     }
 }
 
+/// Helper function. "Derives" `+=` if the value type of a `HashMap` supports it.
+fn hashmap_add_assign<K, V>(lhs: &mut HashMap<K, V>, rhs: HashMap<K, V>)
```

Comment:
There are some deluxe solutions like [this crate I just found](https://docs.rs/merge-hashmap/latest/merge_hashmap/trait.Merge.html), but we don't need all of that. [Haskell](https://hackage.haskell.org/package/containers-0.7/docs/Data-Map-Internal.html#v:unionWith) and [Clojure](https://clojuredocs.org/clojure.core/merge-with) have standard functions to do this.

I think it would be a bit tricky in Rust since you have to work with the ownership semantics, possibly leading to many variants of the function, but it's not impossible.